lib/iris/cube.py

```diff
@@ -1624,17 +1624,31 @@ def shape(self):
 
     @property
     def dtype(self):
-        if self._dtype is None:
-            result = self.core_data.dtype
-        else:
-            result = self._dtype
-        return result
+        dtype = self._dtype
+        if dtype is None:
+            dtype = self.core_data.dtype
+        return dtype
 
     @dtype.setter
     def dtype(self, dtype):
         self._dtype = dtype
 
     @property
+    def fill_value(self):
+        fill_value = self._fill_value
+        if fill_value is not None:
+            try:
+                fill_value = np.asarray([self._fill_value],
+                                        dtype=self.dtype)[0]
+            except OverflowError:
+                fill_value = None
```

Comment:
The `fill_value` requires to be reported within the context of the `dtype`, what ever that is.

We need to catch the case where, for example, a default float `fill_value`, cannot be cast to an `int` when the `dtype` of the cube is changed.

In this case, I've opted to not clearing the `self._fill_value`, so that if the `cube.dtype` change is reverted the "old" `fill_value` is still there. Willing to take advice/thoughts on this ...